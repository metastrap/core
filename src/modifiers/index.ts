export * as nextModifier from './next';
