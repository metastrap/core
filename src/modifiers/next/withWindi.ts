import type { ParseResult } from '@babel/parser';
import type { File } from '@babel/types';
import * as t from '@babel/types';
import { TAstMap, TROAstMap, TObject } from 'types';

export const windiFiles = [
  'windi.config.js',
];

export default function withWindi(
  baseAstMap: TAstMap,
  fullAstMap: TROAstMap,
) {
  /* package.json */
  const pkg = baseAstMap.get('package.json') as TObject;
  pkg.devDependencies = {
    ...pkg.devDependencies || {},
    windicss: '3.5.1',
    'windicss-webpack-plugin': '1.6.7',
  };

  addImportToAppLayout(
    baseAstMap.get('app/layout.jsx') as ParseResult<File>,
    'windi.css',
  );

  nextConfigTransform(
    baseAstMap.get('next.config.js') as ParseResult<File>,
  );

  windiFiles.forEach((file) => {
    baseAstMap.set(file, fullAstMap.get(file) || '');
  });
}

function addImportToAppLayout(
  appLayout: ParseResult<File>,
  importLocation: string,
) {
  const importNodeIndex = appLayout.program
    .body
    .findIndex(
      (node) => node.type === 'ImportDeclaration'
        && node.source.value === importLocation,
    );
  if (!~importNodeIndex) {
    appLayout.program.body.unshift(
      t.importDeclaration(
        [],
        t.stringLiteral(importLocation),
      ),
    );
  }
}

function nextConfigTransform(nextConfig: ParseResult<File>) {
  const { body } = nextConfig.program;

  /* require `@next/mdx` with init */
  body.unshift(
    t.variableDeclaration('const', [
      t.variableDeclarator(
        t.identifier('WindiCSSWebpackPlugin'),
        t.callExpression(
          t.callExpression(
            t.identifier('require'),
            [
              t.stringLiteral('windicss-webpack-plugin'),
            ],
          ),
          [],
        ),
      ),
    ]),
  );

  /* find the line where 'nextConfig' is declared */
  const nextConfigDeclaration = body.find(
    (node) => node.type === 'VariableDeclaration'
      && node.declarations.some(
        (declaration) => (declaration.id as t.Identifier).name === 'nextConfig',
      ),
  ) as t.VariableDeclaration;

  const nextConfigObjectExp = nextConfigDeclaration.declarations.find(
    (declarator) => (declarator.id as t.Identifier).name === 'nextConfig',
  )?.init as t.ObjectExpression;

  /* find the line where 'webpack' is declared */
  const webpackFunction = nextConfigObjectExp.properties.find(
    (property): property is t.ObjectProperty => t.isObjectProperty(property)
      && (property.key as t.Identifier).name === 'webpack',
  ).value as unknown as t.ArrowFunctionExpression;

  /* add windi plugin to webpack config */
  (webpackFunction.body as t.BlockStatement).body.unshift(
    t.expressionStatement(
      t.callExpression(
        t.memberExpression(
          t.memberExpression(
            t.identifier('config'),
            t.identifier('plugins'),
          ),
          t.identifier('push'),
        ),
        [
          t.newExpression(
            t.identifier('WindiCSSWebpackPlugin'),
            [],
          ),
        ],
      ),
    ),
  );
}
