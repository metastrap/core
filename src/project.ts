import JSZip from 'jszip';

import type { TAstValue } from 'types/index';

export function pack(data: Map<string, string>): JSZip {
  const zip = new JSZip();
  data.forEach((content, file) => {
    zip.file(file, content);
  });
  return zip;
}

export function setBaseProject(
  fullProject: Map<string, TAstValue>,
  baseFiles: string[],
): Map<string, TAstValue> {
  const baseProject = new Map<string, TAstValue>();
  fullProject.forEach((content, file) => {
    if (baseFiles.includes(file)) {
      baseProject.set(file, content);
    }
  });
  return baseProject;
}
