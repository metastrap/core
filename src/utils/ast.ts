/* browser */
import { ParseResult, ParserPlugin, parse } from '@babel/parser';
import generate from '@babel/generator';

import type { File } from '@babel/types';

import { TAstMap, TAstValue } from 'types';
import logger from './logger';

enum EFileFeature {
  JAVASCRIPT = 1,
  TYPESCRIPT = 2,
  JSX = 4,
  JSON = 8,
}

export function convertTextToAst(
  fileContentMap: Map<string, string>,
): TAstMap {
  const resultMap = new Map<string, TAstValue>();

  fileContentMap.forEach((content, file) => {
    let featureFlag = 0;
    const extname = file.split('.').pop() || '';
    const babelParsePlugins: ParserPlugin[] = ['classProperties'];

    let result: TAstValue = content;
    switch (extname) {
      case 'js':
        featureFlag |= EFileFeature.JAVASCRIPT;
        break;
      case 'ts':
        featureFlag |= EFileFeature.TYPESCRIPT;
        break;
      case 'jsx':
        featureFlag |= EFileFeature.JSX;
        break;
      case 'tsx':
        featureFlag |= EFileFeature.TYPESCRIPT | EFileFeature.JSX;
        break;
      case 'json':
        try {
          result = JSON.parse(content);
        } catch (e) {
          logger((e as Error).message);
        }
    }

    if (featureFlag & EFileFeature.TYPESCRIPT) babelParsePlugins.push('typescript');
    if (featureFlag & EFileFeature.JSX) babelParsePlugins.push('jsx');

    if (featureFlag & (EFileFeature.JAVASCRIPT | EFileFeature.TYPESCRIPT | EFileFeature.JSX)) {
      try {
        result = parse(content, {
          sourceType: 'module',
          plugins: babelParsePlugins,
        });
      } catch (e) {
        logger((e as Error).message);
        result = content;
      }
    }
    resultMap.set(file, result);
  });
  return resultMap;
}

export function print(ast: Map<string, TAstValue>): Map<string, string> {
  const resultMap = new Map<string, string>();
  ast.forEach((astValue, file) => {
    if (typeof astValue === 'string') {
      resultMap.set(file, astValue);
      return;
    }
    if (astValue.program) {
      resultMap.set(file, generate(astValue as ParseResult<File>).code);
      return;
    }
    if (typeof astValue === 'object') {
      try {
        resultMap.set(file, JSON.stringify(astValue, null, 2));
      } catch (e) {
        resultMap.set(file, astValue.toString?.() || '');
      }
    }
  });
  return resultMap;
}
