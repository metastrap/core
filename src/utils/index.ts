export * as astUtils from './ast';
export * as logger from './logger';
export * as read from './read';
