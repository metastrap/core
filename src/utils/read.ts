import JSZip from 'jszip';

export default async function readZipToMap(
  zip: JSZip,
): Promise<Map<string, string>> {
  const files = Object.keys(zip.files);

  const maybeContentList = files.map(async (file) => {
    if (!zip.files[file].dir) {
      const contents = await zip.files[file].async('text');
      return [
        file,
        contents || '',
      ] as [string, string];
    }
    /* directories will return null */
    return null;
  });

  const contentList = (
    await Promise.all(maybeContentList)
  ).filter((file) => !!file);

  return new Map(contentList);
}
