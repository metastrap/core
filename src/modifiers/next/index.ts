import { withTailwindcss } from './withTailwindcss';
import withMdx from './withMdx';
import withWindi from './withWindi';

/* browser */
export const baseFiles = [
  '.gitignore',
  'app/layout.jsx',
  'app/page.jsx',
  'jsconfig.json',
  'next.config.js',
  'package.json',
];

export const modifiers = {
  withTailwindcss,
  withMdx,
  withWindi,
};
