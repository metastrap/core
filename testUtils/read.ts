// eslint-disable-next-line import/no-extraneous-dependencies
import fg from 'fast-glob';
import path from 'node:path';
import fs from 'node:fs/promises';
import { EFrameworks } from '@/constants/enum';
import { TAstValue } from 'types/index';
import { convertTextToAst } from '@/utils/ast';
import { setBaseProject } from '@/project';
import { baseFiles } from '@/modifiers/next';

export async function readFilesToMap(
  dir: string,
): Promise<Map<string, string>> {
  const files = await fg([`${dir}/**/*`], { dot: true });
  return new Map(
    await Promise.all(
      files.map(async (file) => {
        const contents = await fs.readFile(file, 'utf8');
        return [
          file.replace(`${dir}/`, ''),
          contents || '',
        ] as [string, string];
      }),
    ),
  );
}

export async function readFixture(framework: EFrameworks): Promise<Map<string, string>> {
  return await readFilesToMap(
    path.join(process.cwd(), 'fixtures', framework.toString()),
  );
}

export async function astFromFixture(): Promise<[
  Map<string, TAstValue>,
  Map<string, TAstValue>
]> {
  const fullProject = convertTextToAst(
    await readFilesToMap(
      path.join(process.cwd(), 'fixtures', 'next'),
    ),
  );
  return [
    fullProject,
    setBaseProject(fullProject, baseFiles),
  ];
}
