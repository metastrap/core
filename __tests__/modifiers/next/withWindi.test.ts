import path from 'node:path';
import { TAstValue } from 'types/index';
import { baseFiles } from '@/modifiers/next';
import { convertTextToAst, print } from '@/utils/ast';
import { setBaseProject } from '@/project';
import { readFilesToMap } from 'testUtils/read';
import withWindi, { windiFiles } from '@/modifiers/next/withWindi';

let baseAsTree: Map<string, TAstValue>; let
  fullProject: Map<string, TAstValue>;

describe('modifiers/next/withWindi', () => {
  beforeEach(async () => {
    fullProject = convertTextToAst(
      await readFilesToMap(
        path.join(process.cwd(), 'fixtures', 'next'),
      ),
    );
    baseAsTree = setBaseProject(fullProject, baseFiles);
  });

  test('withWindi', () => {
    withWindi(baseAsTree, fullProject);

    /* ensure `windi.config.js` is added */
    expect(
      windiFiles.map((file) => typeof baseAsTree.get(file)),
    ).toStrictEqual([
      'object',
    ]);

    /* ensure `next.config.js` is modified */
    expect(
      print(baseAsTree).get('next.config.js'),
    ).toMatchSnapshot();

    /* ensure `app/layout.jsx` is modified */
    expect(
      print(baseAsTree).get('app/layout.jsx'),
    ).toMatchSnapshot();
  });
});
