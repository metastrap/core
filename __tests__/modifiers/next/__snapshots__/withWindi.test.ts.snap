// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`modifiers/next/withWindi withWindi 1`] = `
"const WindiCSSWebpackPlugin = require("windicss-webpack-plugin")();
const nextConfig = {
  webpack: config => {
    config.plugins.push(new WindiCSSWebpackPlugin());
    return config;
  }
};
module.exports = nextConfig;"
`;

exports[`modifiers/next/withWindi withWindi 2`] = `
"import "windi.css";
export default function RootLayout({
  children
}) {
  return <html lang="en">
      <body>{children}</body>
    </html>;
}"
`;
